### DISCUSSION TITLE
Custom ThematicGradientColorScheme

### DISCUSSION THREAD

**Discussion by VedankNaik** [iTwin/itwinjs-core#2818] [Q&A]
We are trying to implement **ThematicGradientColorScheme.custom**. We have set **mode=ThematicGradientMode.stepped** and **stepCounts** we are passing through slider value. For each new stepCounts input we are creating new mesh. 
So far from what we've done we are getting mesh based on stepCounts value we are passing initially. But when we change stepCounts value and try to create new mesh it won't display new mesh and still displays the one created with initial stepCount value. 
Inshort it works only on first stepCount value that we send.
What are we missing here?

**Comment by pmconne** [comment]
Please show your code.
@markschlosseratbentley

**Comment by markschlosseratbentley** [comment]
Hi -- You should not have to recreate your mesh - thematic display colorizes whatever is in the view, past or present.
As @pmconne said, can you share your code which updates the thematic display properties?  (Changes the stepCount)

**Reply by VedankNaik** [reply]
public static populateAnalysisStyles(
    mesh: AnalysisMesh,
    steps: number
  ): void {
    const auxdata = mesh.polyface.data.auxData;
    if (!auxdata) return;

    mesh.styles.set("None", undefined);

    auxdata.channels.forEach((channel) => {
      const thematicSettings: ThematicGradientSettingsProps = {};
      thematicSettings.colorScheme = ThematicGradientColorScheme.Custom;
      thematicSettings.mode = ThematicGradientMode.Stepped;
      thematicSettings.stepCount = steps;
      var a = {
        value: 0.16,
        color: ColorDef.create("#FA6B02").toJSON(),
      };
      var b = {
        value: 0.33,
        color: ColorDef.create("#FFFE01").toJSON(),
      };
      var c = {
        value: 0.5,
        color: ColorDef.create("#88FF03").toJSON(),
      };
      var d = {
        value: 0.66,
        color: ColorDef.create("#00ef86").toJSON(),
      };
      var e = {
        value: 0.83,
        color: ColorDef.create("#00d9db").toJSON(),
      };
      var f = {
        value: 1.0,
        color: ColorDef.create("#0069FF").toJSON(),
      };
      thematicSettings.customKeys = [a, b, c, d, e, f];

      const props: AnalysisStyleProps = {
        scalar: {
          channelName: channel.name!,
          range: channel.scalarRange!,
          thematicSettings,
        },
      };

      mesh.styles.set(channel.name!, AnalysisStyle.fromJSON(props));
    });
  }

**Reply by markschlosseratbentley** [reply]
Thanks.  I assumed you were using thematic display such as done in this [sample](https://www.itwinjs.org/sample-showcase/?group=Viewer+Features&sample=thematic-display-sample&imodel=CoffsHarborDemo). You are actually using AnalysisMesh as seen in this [sample](https://www.itwinjs.org/sample-showcase/?group=Viewer+Features&sample=scientific-visualization-sample).
We will check your workflow and get back to you.

**Reply by markschlosseratbentley** [reply]
Hi @VedankNaik - can you share your code which sets the display style's analysisStyle property?  (`viewport.displayStyle.settings.analysisStyle`)  I assume when you change the mesh's stored styles, you also are changing the display style's analysisStyle?

**Reply by VedankNaik** [reply]
public updateMesh(variable: string, steps: number) {
    let newTexture: AnalysisMesh;
    newTexture = contourPlotTexture.createTexture(this.data, steps);
    this.decorator.addMeshes(newTexture);
    this.showDecorator();
    IModelApp.viewManager.selectedView!.displayStyle.settings.analysisStyle =
      this.decorator.meshes.styles.get(variable);
  }

**Reply by markschlosseratbentley** [reply]
Can you show the bigger picture of the flow of the code?  When you update the `stepCount`, what is the flow of code that happens?  Is `populateAnalysisStyles` called, then `updateMesh` called?

**Reply by VedankNaik** [reply]
public  createTexture(data: any, steps: number) {
    const polyface = contourPlotTexture.createPolyface();
    const styles = new Map<string, AnalysisStyle | undefined>();
    const mesh = { polyface, styles };
    contourPlotTexture.populateAnalysisStyles(mesh, steps);
    return mesh;
  }

**Flow: On update stepCount -> updateMesh() -> createTexture() -> populateAnalysisStyles()**

**Reply by markschlosseratbentley** [reply]
I think you are going to want to make sure that `IModelApp.viewManager.selectedView!.displayStyle.settings.analysisStyle =
this.decorator.meshes.styles.get(variable);` happens after your `populateAnalysisStyles` runs.

**Reply by VedankNaik** [reply]
Yes, `IModelApp.viewManager.selectedView!.displayStyle.settings.analysisStyle = this.decorator.meshes.styles.get(variable);` is the last statement that is getting executed in this flow.

**Reply by markschlosseratbentley** [reply]
`IModelApp.viewManager.selectedView!.displayStyle.settings.analysisStyle = this.decorator.meshes.styles.get(variable);` happens in `updateMesh()`, which is called before `populateAnalysisStyles()`, at least based on the flow you provided: Flow: On update stepCount -> updateMesh() -> createTexture() -> populateAnalysisStyles()

Is that not true?

**Reply by VedankNaik** [reply]
Ok flow which I was provided seems incomplete. 
So updateMesh() calls createTexture() which further calls populateAnalysisStyle(). From populateAnalysisStyle() there is return to createTexture(), further createTexture() returns mesh back to updateMesh().
So finally control is back to updateMesh() where  `IModelApp.viewManager.selectedView!.displayStyle.settings.analysisStyle = this.decorator.meshes.styles.get(variable);` is executed at last.

**Reply by markschlosseratbentley** [reply]
Can you verify that the `AnalysisStyle` you are retrieving from `...meshes.styles` and assigning to `IModelApp.viewManager.selectedView!.displayStyle.settings.analysisStyle` has the modified `stepCount` value?

**Reply by VedankNaik** [reply]
Yes, I can see modified `stepCount` value

**Reply by markschlosseratbentley** [reply]
What happens to your old mesh?

**Reply by VedankNaik** [reply]
Old mesh gets replaced with new one

**Reply by markschlosseratbentley** [reply]
@VedankNaik Hello - I set up a test which does a similar workflow (creates a second mesh with an alternative stepCount for the AnalysisStyle), and I am not able to reproduce the problem you are seeing.  I see the stepCount being applied differently to each mesh.

To make further progress, would you be able to provide a small, isolated, working sample which reproduces the issue?

**Reply by pmconne** [reply]
Consider creating [a sandbox](https://www.itwinjs.org/sandboxes/).

**Reply by VedankNaik** [reply]
@markschlosseratbentley Hello, 
To cross verify I did compare data for 2 meshes. As first mesh I took the one which gets created on first run (`stepCount: X`) and second mesh I took which gets created after changing stepCount value (For second mesh I set `stepCount: Y` initially and changed to `stepCount: X`). 
Data used to create meshes in both cases was exactly same with matching `stepCount` values. But visualization in second case does not match with first one.

**Reply by markschlosseratbentley** [reply]
Would you be able to create a simple, small [sandbox](https://www.itwinjs.org/sandboxes/) which demonstrates the issue you are seeing?
I was unable to reproduce the issue (as I understand it) with my local testing.

**Reply by VedankNaik** [reply]
Might not be able to create now, will create whenever possible. 
Also, will it be possible to provide some sample from your side? Thanks.

**Reply by markschlosseratbentley** [reply]
Hi @VedankNaik - I made some progress internally with this problem you reported.  I have reproduced the issue and we are investigating.  We will update you again when we have more information and a possible fix.  Thanks for reporting this.

**Reply by markschlosseratbentley** [reply]
Hi @VedankNaik - We now have a fix for this problem.  The problem would only be apparent on iTwin.js 2.19.  The apparent problem did not occur on iTwin.js 3.0, but only because of another issue which meant gradient textures were not being cached.  Fixes for both of those issues have been pushed to the tip of 3.0.  The fix relevant to 2.19 has just [merged](https://github.com/iTwin/itwinjs-core/pull/2878) (backported) into 2.19.  You can expect the fix to not be present in 2.19.26 (the current 2.19 version), but it should show up in the next one.
